### ISSUE TITLE
Build fails with Boost.ASIO 1.74

### ISSUE THREAD

**Issue by thed636** [yandex/ozo#266]
There is some issue with executors in latest Boost.Asio version. Looks like the problem is that the default executor now is a specialization of `boost::asio::any_executor` template which is incompatible with `boost::asio::executor` as they model the different Executor concepts. Looks like the problem is in `resource_pool` library - https://github.com/elsid/resource_pool/blob/6a4074465cc672fc1999035ef3e851d7c250134c/include/yamail/resource_pool/async/detail/pool_impl.hpp#L105

**Comment by JonasProgrammer** [comment]
Just FYI: The title is a little misleading, as this happens with Boost.ASIO 1.74, regardless of the platform.

I've already [tried to fix this](https://github.com/elsid/resource_pool/compare/master...JonasProgrammer:bug/boost_174_compat), but in the end I failed to get the unit tests running, as the executor interface was completely changed. Maybe it still helps as a starting point.

**Comment by thed636** [comment]
Thanks, Jonas!

Looks like I'm a little bit out of sync with Boost.Asio changes. It looks like there are two Executor concepts - Networking TS Executor and Executors TS Executor. And they are not compatible with each other. I think I should learn more about this situation.

Anyway, thanks for your try. I think this is a good point to start with.

**Comment by JonasProgrammer** [comment]
> Looks like I'm a little bit out of sync with Boost.Asio changes. It looks like there are two Executor concepts - Networking TS Executor and Executors TS Executor. And they are not compatible with each other. I think I should learn more about this situation.

You're right. [The changelog](https://www.boost.org/doc/libs/1_74_0/doc/html/boost_asio/history.html#boost_asio.history.asio_1_18_0___boost_1_74) even mentions a solution regarding the polymorphic executor, but I don't think requiring the user to define `BOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT` in order to use resource_pool/ozo is a solution for the long run. Also, some other things were not compatible with the new executor model, for example `io_context::strand`. Vinnie Falco suggested just using `strand` instead on the cpplang slack, which does work -- but that thing is cumbersome to construct without `make_strand`, the latter being available only from Boost 1.70ish and up.

> Anyway, thanks for your try. I think this is a good point to start with.

Sorry for being of so little help, I just hoped it would be a quick fix, but turns out it wasn't -- and with 1.73 being ok for my purpose, I didn't really want to spend the time to deep-dive into a completely new execution model.

**Comment by thed636** [comment]
Well, looks like we need to define `BOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT` for the time #272  since all the tests that use Executor models need to be rewritten - a lot of work, unfortunately. But we need to build on the Mac OS at the moment.
